lib/dma.c

```diff
@@ -608,34 +610,98 @@ dma_controller_dirty_page_get(dma_controller_t *dma, vfu_dma_addr_t addr,
         return ERROR_INT(EINVAL);
     }
 
-    for (i = 0; i < (size_t)bitmap_size; i++) {
-        uint8_t val = region->dirty_bitmap[i];
-        uint8_t *outp = (uint8_t *)&bitmap[i];
-
-        /*
-         * If no bits are dirty, avoid the atomic exchange. This is obviously
-         * racy, but it's OK: if we miss a dirty bit being set, we'll catch it
-         * the next time around.
-         *
-         * Otherwise, atomically exchange the dirty bits with zero: as we use
-         * atomic or in _dma_mark_dirty(), this cannot lose set bits - we might
-         * miss a bit being set after, but again, we'll catch that next time
-         * around.
-         */
-        if (val == 0) {
-            *outp = 0;
-        } else {
-            uint8_t zero = 0;
-            __atomic_exchange(&region->dirty_bitmap[i], &zero,
-                              outp, __ATOMIC_SEQ_CST);
-        }
+    if (pgsize == dma->dirty_pgsize) {
+        dirty_page_get_same_pgsize(region, bitmap, bitmap_size);
+    } else {
+        dirty_page_get_different_pgsize(region, bitmap, bitmap_size,
+                                        converted_bitmap_size, pgsize,
+                                        dma->dirty_pgsize);
     }
 
 #ifdef DEBUG
-    log_dirty_bitmap(dma->vfu_ctx, region, bitmap, size);
+    log_dirty_bitmap(dma->vfu_ctx, region, bitmap, size, pgsize);
 #endif
 
     return 0;
 }
 
+static void
+dirty_page_exchange(uint8_t *outp, uint8_t *bitmap)
+{
+    /*
+     * If no bits are dirty, avoid the atomic exchange. This is obviously
+     * racy, but it's OK: if we miss a dirty bit being set, we'll catch it
+     * the next time around.
+     *
+     * Otherwise, atomically exchange the dirty bits with zero: as we use
+     * atomic or in _dma_mark_dirty(), this cannot lose set bits - we might
+     * miss a bit being set after, but again, we'll catch that next time
+     * around.
+     */
+    if (*bitmap == 0) {
+        *outp = 0;
+    } else {
+        uint8_t zero = 0;
+        __atomic_exchange(bitmap, &zero, outp, __ATOMIC_SEQ_CST);
+    }
+}
+
+void
+dirty_page_get_same_pgsize(dma_memory_region_t *region, char *bitmap,
+                           size_t bitmap_size)
+{
+    for (size_t i = 0; i < bitmap_size; i++) {
+        dirty_page_exchange((uint8_t *)&bitmap[i], &region->dirty_bitmap[i]);
+    }
+}
+
+void
+dirty_page_get_different_pgsize(dma_memory_region_t *region, char *bitmap,
```

Comment:
I'm trying to understand what needs to be done before I can review the code, let me know if my understanding is correct:

> If the client page size is larger than the server one by N, for each group of N bits of the server, if any bit is set then the bit in the client bitmap must be set (we lose some accuracy).
> If the server page size is smaller than the client one by N, for each server bit that is set then all N bits of that group in the client must be set (we don't lose accuracy).

If the above is correct, wouldn't it be simpler to split the `different_pgsize` into two separate functions?

Reply:
I'm not sure about splitting it into two separate functions altogether but certainly improving variable naming and adding lots of documentation is necessary.